edsnlp/models/nested_ner.py

```diff
@@ -0,0 +1,379 @@
+from enum import Enum
+from typing import Any, Callable, Dict, Iterable, List, Optional, OrderedDict, Tuple
+
+import torch
+from loguru import logger
+from spacy import registry
+from spacy.tokens import Doc
+from thinc.layers import chain
+from thinc.model import Model
+from thinc.shims import PyTorchShim
+from thinc.types import ArgsKwargs, Floats1d, Floats2d, Ints2d
+from thinc.util import (
+    convert_recursive,
+    is_torch_array,
+    is_xp_array,
+    torch2xp,
+    xp2torch,
+)
+
+from edsnlp.models.torch.crf import IMPOSSIBLE, MultiLabelBIOULDecoder
+
+
+class CRFMode(Enum):
+    independent = "independent"
+    joint = "joint"
+    marginal = "marginal"
+
+
+def repeat(t, n, dim, interleave=True):
+    repeat_dim = dim + (1 if interleave else 0)
+    return (
+        t.unsqueeze(repeat_dim)
+        .repeat_interleave(n, repeat_dim)
+        .view(
+            tuple(
+                -1 if (i - dim + t.ndim) % t.ndim == 0 else s
+                for i, s in enumerate(t.shape)
+            )
+        )
+    )
+
+
+def flatten_dim(t, dim, ndim=1):
+    return t.reshape((*t.shape[:dim], -1, *t.shape[dim + 1 + ndim :]))
+
+
+class NestedNERModule(torch.nn.Module):
+    def __init__(
+        self,
+        input_size: Optional[int] = None,
+        n_labels: Optional[int] = None,
+        mode: CRFMode = CRFMode.joint,
+    ):
+        """
+        Nested NER CRF module
+
+        Parameters
+        ----------
+        input_size: int
+            Size of the input embeddings
+        n_labels: int
+            Number of labels predicted by the module
+        mode: CRFMode
+            Loss mode of the CRF
+        """
+        super().__init__()
+
+        self.mode = mode
+
+        assert mode in (CRFMode.independent, CRFMode.joint, CRFMode.marginal)
```

Comment:
I think we might be able to use the Enum constructor for that, although I'm not sure. Or maybe with Pydantic's `validate_arguments`? It's in beta, though.